core/src/main/scala/kafka/tools/DumpLogSegments.scala

```diff
@@ -398,9 +399,34 @@ object DumpLogSegments {
     }
   }
 
-  private class OffsetsMessageParser extends MessageParser[String, String] {
+  // Package private for testing.
+  class OffsetsMessageParser extends MessageParser[String, String] {
+    private val serde = new RecordSerde()
+
     override def parse(record: Record): (Option[String], Option[String]) = {
-      GroupMetadataManager.formatRecordKeyAndValue(record)
+      if (!record.hasKey)
+        throw new RuntimeException(s"Failed to decode message at offset ${record.offset} using offset " +
+          "topic decoder (message had a missing key)")
+
+      try {
+        val r = serde.deserialize(record.key, record.value)
+        (
+          Some(r.key.message.toString),
+          Option(r.value).map(_.message.toString).orElse(Some("<DELETE>"))
```

Comment:
It seems the `r.value` is never null ( https://github.com/apache/kafka/blob/376e9e20dbf7c7aeb6f6f666d47932c445eb6bd1/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/RecordSerde.java#L101 )

Reply:
It could be null. See here: https://github.com/apache/kafka/blob/376e9e20dbf7c7aeb6f6f666d47932c445eb6bd1/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/RecordSerde.java#L92.